src/g_lib/g_lib.cr

```diff
@@ -1,28 +1,63 @@
+require "time"
+
 require "../gobject"
 require_gobject "GLib"
 
 require "./*"
 require "../closure_data_manager"
 
 module GLib
-  def self.idle_add(&block : -> Bool)
+  enum Priority
+    Default      = LibGLib::PRIORITY_DEFAULT
+    Default_Idle = LibGLib::PRIORITY_DEFAULT_IDLE
+    High         = LibGLib::PRIORITY_HIGH
+    High_Idle    = LibGLib::PRIORITY_HIGH_IDLE
+    Low          = LibGLib::PRIORITY_LOW
+  end
+
+  def self.idle_add(priority : Priority = Priority::Default, &block : -> Bool)
     trampoline = LibGLib::SourceFunc.new { block.call ? 1 : 0 }
     LibGLib.idle_add(
-      LibGLib::PRIORITY_DEFAULT_IDLE,
+      priority,
       LibGLib::SourceFunc.new(trampoline.pointer, Pointer(Void).null),
       GObject::ClosureDataManager.register(trampoline.closure_data),
       ->GObject::ClosureDataManager.deregister
     )
   end
 
-  def self.timeout(seconds, &block : -> Bool)
+  def self.timeout_seconds(interval : UInt32, priority : Priority = Priority::Default, &block : -> Bool)
     trampoline = LibGLib::SourceFunc.new { block.call ? 1 : 0 }
+    sourcefunc = LibGLib::SourceFunc.new(trampoline.pointer, Pointer(Void).null)
+    closure = GObject::ClosureDataManager.register(trampoline.closure_data)
     LibGLib.timeout_add_seconds(
-      LibGLib::PRIORITY_DEFAULT_IDLE,
-      UInt32.new(seconds),
-      LibGLib::SourceFunc.new(trampoline.pointer, Pointer(Void).null),
-      GObject::ClosureDataManager.register(trampoline.closure_data),
+      priority,
+      interval,
+      sourcefunc,
+      closure,
+      ->GObject::ClosureDataManager.deregister
+    )
+  end
+
+  def self.timeout_milliseconds(interval : UInt32, priority : Priority = Priority::Default, &block : -> Bool)
+    trampoline = LibGLib::SourceFunc.new { block.call ? 1 : 0 }
+    sourcefunc = LibGLib::SourceFunc.new(trampoline.pointer, Pointer(Void).null)
+    closure = GObject::ClosureDataManager.register(trampoline.closure_data)
+    LibGLib.timeout_add(
+      priority,
+      interval,
+      sourcefunc,
+      closure,
       ->GObject::ClosureDataManager.deregister
     )
   end
+
+  def self.timeout(time : Time::Span, priority : Priority = Priority::Default, &block : -> Bool)
+    if time.to_i == time.total_seconds || time.total_milliseconds > UInt32::Max
+      self.timeout_seconds(time.total_seconds, priority, &block)
+    elsif time.total_milliseconds > 0
+      self.timeout_milliseconds(time.total_milliseconds, priority, &block)
+    else
+      raise ArgumentError.new("Timeout must be at least 1 millisecond")
```

Comment:
Let's move this to `timeout_milliseconds` (`raise ... unless interval > 0`)  and add a similar check to `timeout_seconds actually :)